lib/LANraragi/Model/ReadingGroup.pm

```diff
@@ -0,0 +1,229 @@
+#   If an existing Reading Group ID is supplied, said Reading Group will be updated with the given parameters.
+#   Returns the ID of the created/updated Reading Group.
+sub create_reading_group {
+
+    my ( $name, $rg_id ) = @_;
+    my $redis = LANraragi::Model::Config->get_redis;
+
+    # Set all fields of the group object
+    unless ( length($rg_id) ) {
+        $rg_id = "RG_" . time();
+
+        my $isnewkey = 0;
+        until ($isnewkey) {
+
+            # Check if the group ID exists, move timestamp further if it does
+            if ( $redis->exists($rg_id) ) {
+                $rg_id = "RG_" . ( time() + 1 );
+            } else {
+                $isnewkey = 1;
+            }
+        }
+
+        # Default values for new group
+        $redis->hset( $rg_id, "archives",  "[]" );
+        $redis->hset( $rg_id, "name",   redis_encode($name) );
+    }
+
+    $redis->quit;
+
+    return $rg_id;
+}
+
+# get_reading_group(id, decoded)
+#   Returns the Reading Group matching the given id.
+#   Returns undef if the id doesn't exist.
+sub get_reading_group {
+
+    #my $rg_id = $_[0];
+    my ($rg_id,$decoded) = @_;
+    my $logger = get_logger( "Reading Groups", "lanraragi" );
+    my $redis  = LANraragi::Model::Config->get_redis;
+
+    if ( $rg_id eq "" ) {
+        $logger->debug("No Reading Group ID provided.");
+        return ();
+    }
+
+    unless ( length($rg_id) == 13 && $redis->exists($rg_id) ) {
+        $logger->warn("$rg_id doesn't exist in the database!");
+        return ();
+    }
+
+    my %readinggroup = $redis->hgetall($rg_id);
+
+    # redis-decode the name, and the search terms if they exist
+    ( $_ = redis_decode($_) ) for ( $readinggroup{name});
+
+    if ($decoded) {
+        my @data = get_archive_json_multi(@{ decode_json($readinggroup{archives}) });
+        eval {$readinggroup{archives} = \@data}
+    } else {
+        eval { $readinggroup{archives} = decode_json( $readinggroup{archives} ) };
+    }
+
+    if ($@) {
+        $logger->error("Couldn't deserialize contents of readinggroup $rg_id! $@");
+    }
+
+    # Add the key as well
+    $readinggroup{id} = $rg_id;
+
+    return %readinggroup;
+}
+
+# delete_reading_group(id)
+#   Deletes the reading group with the given ID.
+#   Returns 0 if the given ID isn't a reading group ID, 1 otherwise
+sub delete_reading_group {
+
+    my $rg_id = $_[0];
+    my $logger = get_logger( "Reading Groups", "lanraragi" );
+    my $redis  = LANraragi::Model::Config->get_redis;
+
+    if ( length($rg_id) != 13 ) {
+
+        # Probably not a readinggroup ID
+        $logger->error("$rg_id is not a readinggroup ID, doing nothing.");
+        $redis->quit;
+        return 0;
+    }
+
+    if ( $redis->exists($rg_id) ) {
+        $redis->del($rg_id);
+        $redis->quit;
+        return 1;
+    } else {
+        $logger->warn("$rg_id doesn't exist in the database!");
+        $redis->quit;
+        return 1;
+    }
+}
+
+# add_to_reading_group(categoryid, arcid)
```

Comment:
Yep, because reordering is easier in the Frontend is that I made it this way, but an "Insert one" is maybe at least necessary, because that way you can insert easily a file from the search by getting only the id of the group instead of getting it's whole content(Although, I think that to get the name of the group I'm already returning it's content list).

Reply:
If we use sorted sets for the database structure it'd probably be easier to have insert/remove API endpoints at least -- No need to serialize/deserialize json every time 🥲